LAMA to Dask: updates to fix CI workflow setup errors

sadielbartholomew opened the pull request:
The Actions CI workflows have recently been producing jobs that all fail early (or are cancelled early due to parallel such failures) on the same setup errors, as copied in summary below. This PR aims to test, and then merge, a fix for this, so we can move onto seeing a true pass or fail output.

Despite what the note on the penultimate quoted line says, I am fairly certain this *is* actually a pip-related issue stemming from older versions of `pip` being used because (I was unaware of this before, but after investigation here I now realise) we are pretty behind in terms of versions specified for the so-called 'reusable' workflows which we specify in order to do common Actions jobs tasks such as checking out a repository and setting up Python, etc. By updating to the latest versions of those, notably the latter, I think we will get a `pip` version which won't hit this error. Testing that theory now here via the set of job 'checks' that get run on this PR...

(Whether the version updates fix this issue at hand or not, it is worthwhile to do the update. I'll go in and update all of the other workflow files in a similar way shortly in a separate PR.)


### Log output outlining the setup error in question

```console
Run cd /home/runner/work/cf-python/cf-python/cfdm
Obtaining file:///home/runner/work/cf-python/cf-python/cfdm
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Checking if build backend supports build_editable: started
  Checking if build backend supports build_editable: finished with status 'done'
  Getting requirements to build editable: started
  Getting requirements to build editable: finished with status 'done'
  Preparing editable metadata (pyproject.toml): started
  Preparing editable metadata (pyproject.toml): finished with status 'done'
Collecting netcdf-flattener>=1.2.0
  Downloading netcdf-flattener-1.2.0.tar.gz (26 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Requirement already satisfied: packaging>=20.0 in /usr/share/miniconda3/envs/cf-latest/lib/python3.8/site-packages (from cfdm==1.10.0.1) (21.3)
Requirement already satisfied: numpy>=1.15 in /usr/share/miniconda3/envs/cf-latest/lib/python3.8/site-packages (from cfdm==1.10.0.1) (1.23.5)
Collecting netCDF4>=1.5.4
  Downloading netCDF4-1.6.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 92.8 MB/s eta 0:00:00
Collecting cftime>=1.6.0
  Downloading cftime-1.6.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB [12](https://github.com/NCAS-CMS/cf-python/actions/runs/3591398929/jobs/6045870873#step:9:13)7.2 MB/s eta 0:00:00
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/share/miniconda3/envs/cf-latest/lib/python3.8/site-packages (from packaging>=20.0->cfdm==1.10.0.1) (3.0.9)
Building wheels for collected packages: cfdm, netcdf-flattener
  Building editable for cfdm (pyproject.toml): started
  Building editable for cfdm (pyproject.toml): finished with status 'done'
  Created wheel for cfdm: filename=cfdm-1.10.0.1-0.editable-py3-none-any.whl size=6830 sha256=1d778e6304175f3a884dcdd7735f3989e0d48e1c49484c935416a9ac42a80a91
  Stored in directory: /tmp/pip-ephem-wheel-cache-cjecr_x_/wheels/33/57/e7/a03234108b4035cfc4884dbaf64835d61f145f5e091c1b048b
  Building wheel for netcdf-flattener (setup.py): started
  Building wheel for netcdf-flattener (setup.py): finished with status 'done'
  Created wheel for netcdf-flattener: filename=netcdf_flattener-1.2.0-py3-none-any.whl size=16602 sha256=115[13](https://github.com/NCAS-CMS/cf-python/actions/runs/3591398929/jobs/6045870873#step:9:14)d772a67d3f845da3a8b28d5b570802288b52ce1d2a62b2fe899cb4dd7ee
  Stored in directory: /home/runner/.cache/pip/wheels/3a/5c/01/ecdde7cba02b1abe2ecd807612f4ca9e39927db21e40cb5c6a
Successfully built cfdm netcdf-flattener
Installing collected packages: cftime, netCDF4, netcdf-flattener, cfdm
Successfully installed cfdm-1.10.0.1 cftime-1.6.2 netCDF4-1.6.2 netcdf-flattener-1.2.0
Obtaining file:///home/runner/work/cf-python/cf-python/main
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Checking if build backend supports build_editable: started
  Checking if build backend supports build_editable: finished with status 'done'
  Getting requirements to build editable: started
  Getting requirements to build editable: finished with status 'error'
  error: subprocess-exited-with-error
  
  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      running egg_info
      creating cf_python.egg-info
      writing cf_python.egg-info/PKG-INFO
      writing dependency_links to cf_python.egg-info/dependency_links.txt
      writing requirements to cf_python.egg-info/requires.txt
      writing top-level names to cf_python.egg-info/top_level.txt
      writing manifest file 'cf_python.egg-info/SOURCES.txt'
      error: package directory 'cf/data/abstract' does not exist
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
Error: Process completed with exit code 1.
```

sadielbartholomew commented:
> By updating to the latest versions of those, notably the latter, I think we will get a pip version which won't hit this error. Testing that theory now here via the set of job 'checks' that get run on this PR...

OK the results are in and my theory was wrong :pensive: . The version updates need doing anyway so will stay, but something else is needed to fix the underlying issue. I'll have a look tomorrow, probably first I'll try adding a command to upgrade `pip` at the appropriate place and see if that helps...

davidhassell commented:
Thanks for investigating this, Sadie - it'll be great to get these working again.

sadielbartholomew commented:
Opening and closing to re-trigger the jobs...

sadielbartholomew commented:
Ang again, last time there was nothing that would cause any difference, just some informational print-outs, this time I am upgrading `pip` which might help (if it can be upgraded at all)...

sadielbartholomew commented:
And again, last time there was nothing that would cause any difference, just some informational print-outs, this time I am upgrading pip which might help (if it can be upgraded at all)...

sadielbartholomew commented:
Argh, I should have just listened to the note "note: This error originates from a subprocess, and is likely not a problem with pip." in the first place! The issue was that an old module, now moved elsewhere, was specified in the `setup.py` and causing the failure (I thought `error: package directory 'cf/data/abstract' does not exist` was a symptom and not referring to the root cause of the issue, given that editable installation has always been working fine for me locally, and I could see no reference to `data.abstract` in any `__init__.py`, but forgot to check the `setup.py` and it was indeed the problem :face_with_head_bandage: ).

Anyway, re-triggering to show that the setup stages don't trip on the same error now, and to see how far the jobs get now...

sadielbartholomew commented:
I squashed a few commits so had to do a force-push just now, hence the many commits being reported at once.

I've updated all of the workflows to reference the latest 'reusable workflows' in [23bfcf9](https://github.com/NCAS-CMS/cf-python/pull/520/commits/23bfcf902298ecbead5a6865647e9be000a23a51). Those workflows don't get run as jobs on this branch, though I am thinking that we can shortly start using the old standard full test-suite run since we are aspiring, and not far off, getting the full test suite to pass, as outlined in #428. Then I can just delete the `dask-migration-testing.yml` temporary workflow as it is no longer needed. But I'll wait until we get that (essentially `test_Data`) to pass, because we are seeing some issues due to undetected files (see #495).

I'll just add a final commit to convert the other workflows to do a no-dependency install of `cf`, which is more robust, then re-confirm the job results, then I can merge this as the underlying issue has been fixed.

sadielbartholomew commented on setup.py:
```diff
@@ -306,7 +306,6 @@ def compile():
         "cf",
         "cf.mixin",
         "cf.data",
-        "cf.data.abstract",
```
@davidhassell what do you want me to do here, namely to edit as packages in the `setup.py`, or should we do it later / elsewhere (see my comment above)?

sadielbartholomew replied:
You're probably offline now, so I'll take your immediate suggestion and add `"cf.data.array.abstract"` for now so I can merge, but we will need to ensure we've covered everything in the `packages` sometime before we release...

sadielbartholomew commented:
Re-triggering again with an open-close...

sadielbartholomew commented:
Again...

sadielbartholomew commented:
Setup still completes successfully but we're now hitting an issue relating to #524, which I will investigate separately. But the issue behind the PR has been solved, namely to fix the setup issues, and I have updated the various Actions workflows in the process, by:

* updating the `setup.py` for new modules;
* using the latest versions for all referenced and utilised reusable workflows;
* moving to use a no-dependency install of `cf` which is more robust.

Doing one final re-trigger check and we are ready to merge...

sadielbartholomew commented:
Setup still completes successfully but we're now hitting an issue relating to https://github.com/NCAS-CMS/cf-python/pull/524, which I will investigate separately. But the issue behind the PR has been solved, namely to fix the setup issues, and I have updated the various Actions workflows in the process, by:

* updating the setup.py for new modules;
* using the latest versions for all referenced and utilised reusable workflows;
* moving to use a no-dependency install of cf which is more robust.

Doing one final re-trigger check and we are ready to merge...

sadielbartholomew commented:
Same results as before, as expected. Merging...
